web/src/components/storage/VolumeForm.jsx

```diff
@@ -721,13 +722,14 @@ export default function VolumeForm({ id, volume: currentVolume, templates = [],
   return (
     <Form id={id} onSubmit={submitForm}>
       <FormGroup isRequired label={_("Mount point")} fieldId="mountPoint">
-        <MountPointFormSelect
+        {currentVolume !== undefined} && <p> {state.formData.mountPoint} </p>
```

Comment:
I didn't check it by myself yet, but reading the code I'd say this is incomplete because now nothing will be shown when _currentVolume_ is defined. Remember that the form is used for both cases, when defined and when not.


```jsx
{currentVolume !== undefined && <p>{state.formData.mountPoint</p>}
```

Note the use of less curly braces. You can have a look to React conditional rendering documentation at https://react.dev/learn/conditional-rendering.

IMHO, a bit more understandable if you use the Agama `core/<If>` component instead,

```jsx
<If condition={currentVolume !== undefined} then={<p>{state.formData.mounPoint</p>} />
```


Moreover, actually for this use case you need to use both `core/If` params, `then` and `else` in order to show the selector or the mount point depending on the value of _currentVolume_. Something like, 


```jsx
const MountPointSelector = () => (
  <MountPointFormSelect
    ...
  />
);

const MountPoint = () => <p>{state.formData.mountPoint</p>;

<If
  condition={currentVolume}
  then={<MountPointSelector />
  else={<MountPoint />
 /> 
```

Please, note that code above is just an incomplete example for illustrating how I'd proceed. Note that I'm using kind of internal components instead of just variables in order to not evaluate them if not needed. Check the React documentation about components at https://react.dev/learn/your-first-component and also have a look to https://javascript.info/arrow-functions-basics if you're not familiar with arrow functions to fully understand the example.

Reply:
Hello Sir, Thank you so much for the detailed explanation of the errors I have made and also for the Recommendations to follow for learning more about the project.  I will try my best to rectify my mistakes.
Sorry for the inconvenience caused.

Reply:
No problem at all.

The only mistake there was not mounting/rendering the selector when needed. That's why unit tests are really important, for covering such use cases and checking that we do not overlook them. The rest was just a different way of thinking/doing things.

In any case, if you tell us about your knowledge level, maybe we can find easier tasks that allow you to learn and contribute at the same time. Also, I'd like to know if you finally were able to run the environment in the virtual machine or not. It's important that you can safely play with the project while contributing to help you to understand what is going on.

Of course, these links I put in other links are always helpful no matter your level or the project you're contributing (as long as the project is about JavaScript / React).

Reply:
I am a beginner to open source and am currently in my 2nd year of university.
My skills Include C/C++, JavaScript, Typescript(Basic), Python(Basic), MERN stack (still practicing), Next JS (Basic).
I am still having issues with running the environment. I did as sir @joseivanlopez recommended https://github.com/openSUSE/agama/issues/932#issuecomment-1863007573
but when I point to localhost 8080 It's not working.
![Screenshot 2024-01-02 180655](https://github.com/openSUSE/agama/assets/125825782/d110b4cf-2ef8-4a78-895e-58301dc4fabc)
I selected openSUSE Tumbleweed when asked and also installed all the packages. After that, it told me to reboot to start with the project but when I did reboot it was still showing this page only.
Thank you for the response.

Reply:
I am thinking of running the project locally. Is it okay to do so.

Reply:
> I selected openSUSE Tumbleweed when asked and also installed all the packages. After that, it told me to reboot to start with the project but when I did reboot it was still showing this page only.

Interesting. Can you explain this a bit more, please? I'm not sure if I got it. Did you hit the `Install` button after selecting `openSUSE Tumbleweed`?

Reply:
Yes, sir, I did.
I will attach some screenshots of the issues in 10 minutes as my internet is not so good.

Reply:
> I am thinking of running the project locally. Is it okay to do so.

At this moment, I **do not** think so. Maybe I'm wrong, but I suspect you have not clear the implications of hitting the `Install` button: as the UI states, **it will format the selected device**. That's why we recommend using a VM for development, to keep your data in safe.

Just out of curiosity, do you really know what this project is about? Sorry if you feel the question does not proceed, but I prefer to have pretty clear your knowledge about Agama.

Reply:
> Yes, sir, I did. I will attach some screenshots of the issues in 10 minutes as my internet is not so good.

Then It was working as expected.

* You created the VM
* Used the Agama Live ISO
* Reached the Agama installer
* Selected a product and installed it

The thing didn't work was opening Agama _remotely_ from the host by opening http://localhost:8080 in your browser. Right?

Reply:
I have read the agama README page and from what I understand It is a tool that makes the installation process of linux easier.

Reply:
> > Yes, sir, I did. I will attach some screenshots of the issues in 10 minutes as my internet is not so good.
> 
> Then It was working as expected.
> 
> * You created the VM
> * Used the Agama Live ISO
> * Reached the Agama installer
> * Selected a product and installed it
> 
> The thing didn't work was opening Agama _remotely_ from the host by opening http://localhost:8080 in your browser. Right?

Yes, sir.
But I also don't exactly understand why after rebooting it takes me back to the same installation page

Reply:
> But I also don't exactly understand why after rebooting it takes me back to the same installation page

I guess it's because your virtual machine still booting from the live ISO. If so, it's the expected behavior. Try to _unmount_ the live ISO after install and before rebooting. Your virtual machine should take you to an openSUSE Tumbleweed system.

Reply:
> > But I also don't exactly understand why after rebooting it takes me back to the same installation page
> 
> I guess it's because your virtual machine still booting from the live ISO. If so, it's the expected behavior. Try to _unmount_ the live ISO after install and before rebooting. Your virtual machine should take you to an openSUSE Tumbleweed system.

Okay sir I will try that.

Reply:
> > Yes, sir, I did. I will attach some screenshots of the issues in 10 minutes as my internet is not so good.
> 
> Then It was working as expected.
> 
> * You created the VM
> * Used the Agama Live ISO
> * Reached the Agama installer
> * Selected a product and installed it
> 
> The thing didn't work was opening Agama _remotely_ from the host by opening http://localhost:8080 in your browser. Right?

when I ran ip addr show in my VM terminal It showed 2 ip addresses 127.0.0.1 and 10.0.2.255
I tried both the ip addresses for running the " COCKPIT_TARGET=<virtual-machine-ip> npm run server".
I wanted to ask could there have been an error here.
Thank you.

Reply:
The `127.0.0.1` IP is the local host, i.e. the machine from which you are connecting, that obviously cannot work.
The `10.0.2.255` is a broadcast address (they end with .255), it is not a normal IP address.

Run `ip address show scope global` in the VM, the IP to use is the one following `inet` label.

Reply:
BTW we plan to use mDNS/Zeroconf in the future, so instead of the IP you could use "agama.local" host name (or something like that). That will be much easier than figuring out the correct IP to use.

Reply:
> The `127.0.0.1` IP is the local host, i.e. the machine from which you are connecting, that obviously cannot work. The `10.0.2.255` is a broadcast address (they end with .255), it is not a normal IP address.
> 
> Run `ip address show scope global` in the VM, the IP to use is the one following `inet` label.

@lslezak sir, Thank you so much for the clarification.

Reply:
> > But I also don't exactly understand why after rebooting it takes me back to the same installation page
> 
> I guess it's because your virtual machine still booting from the live ISO. If so, it's the expected behavior. Try to _unmount_ the live ISO after install and before rebooting. Your virtual machine should take you to an openSUSE Tumbleweed system.

If you want to boot the installed system then you have to manually select the "Boot from Hard Disk" option at the bottom, otherwise it will boot the installation system again.

![agama-boot-menu](https://github.com/openSUSE/agama/assets/907998/031c7d72-2c9c-4de6-8421-9e7b7ba27ccc)

Alternatively you can change the VM configuration to boot from hard disk instead of DVD.

Note: Unfortunately the bootloader configuration on the LiveCD is read-only so we cannot change the default option after the installation is finished.

Reply:
@lslezak sir, Thank you so much for the response. I will surely try that and will let you know.

Reply:
Hello sir, Sorry for the constant trouble. As I have mentioned the live ISO file was successfully booted in the VM and I logged in there then I booted the web UI as sir @joseivanlopez instructed https://github.com/openSUSE/agama/issues/932#issuecomment-1863007573 but unfortunately, it is still causing an issue when connecting to the COCKPIT Server
![Screenshot 2024-01-03 221932](https://github.com/openSUSE/agama/assets/125825782/0cb652ea-1eb6-491d-9a5d-34dc8f505d3c)
![Screenshot 2024-01-03 222101](https://github.com/openSUSE/agama/assets/125825782/1888d905-4a62-4b6e-ac75-2881292ae037)
![Screenshot 2024-01-03 222130](https://github.com/openSUSE/agama/assets/125825782/1ebb35d5-f994-4a18-b86b-6aa383e17bd9)
![Screenshot 2024-01-03 223202](https://github.com/openSUSE/agama/assets/125825782/3d310967-1ae1-443b-aa5d-e12df6e4697c)
I would really appreciate the help, Thank you.

Reply:
Hi @nagu165, 

Lookint at your screenshots everything looks fine, except that you are running the installed product not the Agama live ISO. I mean, by the latest screenshot I guess that your virtual machine booted from the hard disk and logged into a Linux system, most probably a product offered and installed by Agama.

Think that here we're managing two different, but related, concepts:

* The Agama live ISO, which allows you to install a product and/or use it as part of your Agama development environment by making use of the running cockpit and Agama services there.
* The installed product by Agama, which is an (open)SUSE linux distribution that by default does not have either, cockpit nor Agama services installed and running.

I believe that you started the Agama live ISO and then installed a product and booted the machine from the virtual hard disk. If so, as said above, cockpit and Agama services just gone, and that's why you got the "Cannot connect to the Cockpit server" message.

Summarizing, while working in the Agama development your virtual machine should be running the Agama live ISO. At least for now.

Please, let us know if it is clear now or there is something you still not understanding.

Reply:
> Hi @nagu165,
> 
> Lookint at your screenshots everything looks fine, except that you are running the installed product not the Agama live ISO. I mean, by the latest screenshot I guess that your virtual machine booted from the hard disk and logged into a Linux system, most probably a product offered and installed by Agama.
> 
> Think that here we're managing two different, but related, concepts:
> 
> * The Agama live ISO, which allows you to install a product and/or use it as part of your Agama development environment by making use of the running cockpit and Agama services there.
> * The installed product by Agama, which is an (open)SUSE linux distribution that by default does not have either, cockpit nor Agama services installed and running.
> 
> I believe that you started the Agama live ISO and then installed a product and booted the machine from the virtual hard disk. If so, as said above, cockpit and Agama services just gone, and that's why you got the "Cannot connect to the Cockpit server" message.
> 
> Summarizing, while working in the Agama development your virtual machine should be running the Agama live ISO. At least for now.
> 
> Please, let us know if it is clear now or there is something you still not understanding.

Hello Sir, Thank you so much for the response. From what I understand I have installed the openSUSE Linux distribution (Tumbleweed) from the installer provided by agama then as you have mentioned in https://github.com/openSUSE/agama/pull/973#discussion_r1439427331 I unmounted the live iso to start the Tumbleweed CLI right?. After reading your response I have tried running the Agama Live ISO again but it still showed cockpit error. Can you please tell me whether I am supposed to install a product here or not.
![Screenshot 2024-01-02 180655](https://github.com/openSUSE/agama/assets/125825782/ac3f9da4-5156-45b4-b872-4086c3a6b658)
Also can you tell me what I should expect when running the agama Live ISO so that I will know when and where to stop and try again differently.
Thank you so much for your help as I am new to this ( I used linux before but this feels quite different than what I experienced with other linux distros ) and am learning a lot thanks to you.

Reply:
> Hello Sir, Thank you so much for the response. From what I understand I have installed the openSUSE Linux distribution (Tumbleweed) from the installer provided by agama then as you have mentioned in [#973 (comment)](https://github.com/openSUSE/agama/pull/973#discussion_r1439427331) I unmounted the live iso to start the Tumbleweed CLI right?. 

I really sorry for the misunderstanding. In such a comment, I was just trying to explain why you got back to the installer once the installation finished and you rebooted the machine. I thought you were interested on booting the installed product just out of curiosity.

> After reading your response I have tried running the Agama Live ISO again but it still showed cockpit error. Can you please tell me whether I am supposed to install a product here or not. 

Please, check the IP of the virtual machine when running the live ISO and make sure it matches the value you are setting via COCKPIT_TARGET.

As said above, here we're mixing two concepts:

* The installer (Agama)
* The installed distribution (openSUSE Tumbleweed in your example)

For the installer development purpose, you need the first one, Agama. While a product (openSUSE) can be set for running the installer on top of it, let's start simple and say that all you need by now if the Agama live ISO running.

So, let's recap the steps mentioned at https://github.com/openSUSE/agama/issues/932#issuecomment-1863007573

* Download the Agama live ISO :heavy_check_mark: 
* Set a virtual machine for running the Agama live ISO :heavy_check_mark: 
* Clone the openSUSE/agama repo on your machine :heavy_check_mark: 
* Run `COCKPIT_TARGET=<IP> npm run server` from the agama/web directory :heavy_check_mark: 
* Goes to `http://localhost:8080` :heavy_check_mark: 
* Check that the installer is running :heavy_check_mark: 
* Check that you can play around with the installer :heavy_multiplication_x: 
* Make changes in the source code to see them in the installer :heavy_multiplication_x: 

Since you are getting a Cockpit error, please check the VM ip.


> Thank you so much for your help

My pleasure

> as I am new to this ( I used linux before but this feels quite different than what I experienced with other linux distros ) and am learning a lot thanks to you.

Glad to know Agama project is helping you to learn new things. Great :clap: :clap: :clap: About the difference you found is because mixed concepts: installer vs installed system.

Let's see if I have been able to explain it a bit better.

Reply:
I think I found the problem after seeing the `10.0.2.15` IP address.

Are you using VirtualBox for running the virtual machine (VM)? That `10.0.2.15` IP address is used in VirtualBox VM as the default IP when using the NAT network type. And in that mode the machine is not reachable from outside, not even from the host (the same  machine) unless you configure port forwarding.

Configuring and using port forwarding is not trivial so I'd recommend to use the bridged network mode instead. Just go to the VM configuration and change the "NAT" option to the "Bridged Adapter":

![vbox_network](https://github.com/openSUSE/agama/assets/907998/94eb42e3-bf81-4a09-a6a0-672ae45e56d8)

(Note: You should change that only if the VM is not running, if it is running you should shut it down first.)

You can find more details about the network modes in the [VirtualBox documentation](https://www.virtualbox.org/manual/ch06.html#networkingmodes).

Reply:
Thanks a lot for that hing @lslezak!

Let's see if @nagu165 is able to get the development environment ready after applying it :crossed_fingers:

Reply:
Thank you so much @dgdavid sir and @lslezak sir. After following your instructions I was successful in setting up the development environment. I will do my best to contribute well to the project.

Reply:
Glad to read that ;)

We've learned a few things about problems faced by new contributors. We hope to update the development documentation ASAP.

BTW, a bit off-topic but now that we know your experience, I'm curious to know a bit more about how @balsa-asanovic set up his development environment since he already made a few contributions. If you don't mind @balsa-asanovic, tell us a bit about your environment and if you remember you had troubles when setting up Agama. Thanks in advance.

Reply:
Hey guys,

@nagu165 welcome to the project!

>  I'm curious to know a bit more about how @balsa-asanovic set up his development environment

I use two environments, initially I made a separate partition on my HDD and installed Open SUSE Tumbleweed. Then I'd boot to Tumbleweed and proceed with the development there.

As to the project itself, I'd just do a git clone and then run ./setup.sh (with sudo).
I don't remember having any major issues. I do remember running into some errors when running the ./setup.sh for the first couple of times, but this usually amounted to doing `zypper install somePackage` that was missing.

For example, the last issue I had was actually just a couple days ago. After syncing my fork with the main project the ./setup.sh starting reporting error for the rustc version. I ended up updating cargo (and rustc within that) and this fixed the issue.

Lately, I've been using a second environment. I've set up a Windows Subsystem for Linux on my Windows OS. I installed Tumbleweed in WSL as well. I've used this WSL for the large part of my last PR, so I can say that it is working nicely.

Previously, I did encounter that some things didn't work as expected in Agama application when ran in WSL and then I'd just revert to using the Tumbleweed on my HDD. However, lately everything seems to work as expected in WSL as well, this is possibly due to the improvements in the project itself.

In relation the issues @nagu165 is having, I didn't really use VirtualBox or VmWare at all, so not much help there. 😅